File: problem/fasta/solution/php/3.php
<?php
/* The Computer Language Benchmarks Game
   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

   contributed by Wing-Chung Leung
   modified by Isaac Gouy
   modified by anon
 */

ob_implicit_flush(1);
ob_start(NULL, 4096);

$last = 42.0;
function gen_random(&$last, &$randoms, $max = 1.0, $ia = 3877.0, $ic = 29573.0,
$im = 139968.0) {
   foreach($randoms as &$r) {
      $r = $max * ($last = ($last * $ia + $ic) % $im) / $im;
   }
}

/* Weighted selection from alphabet */

function makeCumulative(&$genelist) {
   $cumul = 0.0;
   foreach($genelist as $k=>&$v) {
      $cumul = $v += $cumul;
   }
}


/* Generate and write FASTA format */

function makeRandomFasta(&$genelist, $n) {
   $width = 60;
   $lines = (int) ($n / $width);
   $pick = str_repeat(ʼ?ʼ, $width)."\n";
   $randoms = array_fill(0, $width, 0.0);
   global $last;

   // full lines
   for ($i = 0; $i < $lines; ++$i) {
      gen_random($last, $randoms);
      $j = 0;
      foreach ($randoms as $r) {
         foreach($genelist as $k=>$v) {
            if ($r < $v) {
               break;
            }
         }
         $pick[$j++] = $k;
      }
      echo $pick;
   }

   // last, partial line
   $w = $n % $width;
   if ($w !== 0) {
      $randoms = array_fill(0, $w, 0.0);
      gen_random($last, $randoms);
      $j = 0;
      foreach ($randoms as $r) {
         foreach($genelist as $k=>$v) {
            if ($r < $v) {
               break;
            }
         }
         $pick[$j++] = $k;
      }
      $pick[$w] = "\n";
      echo substr($pick, 0, $w+1);
   }

}


function makeRepeatFasta($s, $n) {
   $i = 0; $sLength = strlen($s); $lineLength = 60;
   while ($n > 0) {
      if ($n < $lineLength) $lineLength = $n;
      if ($i + $lineLength < $sLength){
         print(substr($s,$i,$lineLength)); print("\n");
         $i += $lineLength;
      } else {
         print(substr($s,$i));
         $i = $lineLength - ($sLength - $i);
         print(substr($s,0,$i)); print("\n");
      }
      $n -= $lineLength;
   }
}


/* Main -- define alphabets, make 3 fragments */

$iub=array(
   ʼaʼ => 0.27,
   ʼcʼ => 0.12,
   ʼgʼ => 0.12,
   ʼtʼ => 0.27,

   ʼBʼ => 0.02,
   ʼDʼ => 0.02,
   ʼHʼ => 0.02,
   ʼKʼ => 0.02,
   ʼMʼ => 0.02,
   ʼNʼ => 0.02,
   ʼRʼ => 0.02,
   ʼSʼ => 0.02,
   ʼVʼ => 0.02,
   ʼWʼ => 0.02,
   ʼYʼ => 0.02
);

$homosapiens = array(
   ʼaʼ => 0.3029549426680,
   ʼcʼ => 0.1979883004921,
   ʼgʼ => 0.1975473066391,
   ʼtʼ => 0.3015094502008
);

$alu =
   ʼGGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGʼ .
   ʼGAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGAʼ .
   ʼCCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAATʼ .
   ʼACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAʼ .
   ʼGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGGʼ .
   ʼAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCʼ .
   ʼAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAAʼ;

$n = 1000;

if ($_SERVER[ʼargcʼ] > 1) $n = $_SERVER[ʼargvʼ][1];

makeCumulative($iub);
makeCumulative($homosapiens);

echo ">ONE Homo sapiens alu\n";
makeRepeatFasta($alu, $n*2);

echo ">TWO IUB ambiguity codes\n";
makeRandomFasta($iub, $n*3);

echo ">THREE Homo sapiens frequency\n";
makeRandomFasta($homosapiens, $n*5);

File: problem/fasta/solution/ocaml/3.ml
(* The Computer Language Benchmarks Game
 * https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
 *
 * contributed by Troestler Christophe
 * modified by Mauricio Fernandez
 * changed use of Strings to Bytes where mutation is needed, by Tony Tavener
 *)

(* Random number generator *)
let im = 139968
and ia = 3877
and ic = 29573

let last = ref 42 and im_f = float im
let gen_random  max =
  let n = (!last * ia + ic) mod im in
    last := n;
    max *. float n /. im_f

module Cumul_tbl =
struct
  type t = { probs : float array; chars : char array }

  let make a = let p = ref 0.0 in
    {
      probs = Array.map (fun (_, p1) -> p := !p +. p1; !p) a;
      chars = Array.map fst a;
    }

  let rand_char t =
    let p = gen_random 1.0 in
    let i = ref 0 and ps = t.probs in
      while p >= ps.(!i) do incr i done;
      t.chars.(!i)
end

let width = 60

let make_random_fasta id desc table n =
  Printf.printf ">%s %s\n" id desc;
  let table = Cumul_tbl.make table in
  let line = Bytes.make (width+1) ʼ\nʼ in
  for i = 1 to n / width do
    for j = 0 to width - 1 do Bytes.set line j (Cumul_tbl.rand_char table) done;
    print_bytes line;
  done;
  let w = n mod width in
  if w > 0 then (
    for j = 1 to w do print_char(Cumul_tbl.rand_char table); done;
    print_char ʼ\nʼ
  )

(* [write s i0 l w] outputs [w] chars of [s.[0 .. l]], followed by a
   newline, starting with [s.[i0]] and considering the substring [s.[0
   .. l]] as a "circle".
   One assumes [0 <= i0 <= l <= String.length s].
   @return [i0] needed for subsequent writes.  *)
let rec write s i0 l w =
  let len = l - i0 in
  if w <= len then (output_substring stdout s i0 w; print_char ʼ\nʼ; i0 + w)
  else (output_substring stdout s i0 len; write s 0 l (w - len))

let make_repeat_fasta id desc src n =
  Printf.printf ">%s %s\n" id desc;
  let l = String.length src
  and i0 = ref 0 in
  for i = 1 to n / width do
    i0 := write src !i0 l width;
  done;
  let w = n mod width in
  if w > 0 then ignore(write src !i0 l w)


let alu = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGG\
GAGGCCGAGGCGGGCGGATCACCTGAGGTCAGGAGTTCGAGA\
CCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAAT\
ACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCA\
GCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGG\
AGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCC\
AGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA"

let iub = [| (ʼaʼ, 0.27);  (ʼcʼ, 0.12);  (ʼgʼ, 0.12);  (ʼtʼ, 0.27);
             (ʼBʼ, 0.02);  (ʼDʼ, 0.02);  (ʼHʼ, 0.02);  (ʼKʼ, 0.02);
             (ʼMʼ, 0.02);  (ʼNʼ, 0.02);  (ʼRʼ, 0.02);  (ʼSʼ, 0.02);
             (ʼVʼ, 0.02);  (ʼWʼ, 0.02);  (ʼYʼ, 0.02);  |]

let homosapiens = [| (ʼaʼ, 0.3029549426680);    (ʼcʼ, 0.1979883004921);
                     (ʼgʼ, 0.1975473066391);    (ʼtʼ, 0.3015094502008);  |]

let () =
  let n = try int_of_string(Array.get Sys.argv 1) with _ -> 1000 in
  make_repeat_fasta "ONE" "Homo sapiens alu" alu (n*2);
  make_random_fasta "TWO" "IUB ambiguity codes" iub (n*3);
  make_random_fasta "THREE" "Homo sapiens frequency" homosapiens (n*5)


